### PR TITLE
[ROCm] Update ROCm XLA backend to use GCN Arch Name 

### PR DESCRIPTION
**Pull request by deven-amd**
Starting with ROCm 4.1(?),  TF XLA will need to correctly populate the "target-feature" when creating the LLVM AMDGPUTarget, in order to get optimal performance. The "target-feature"s supported by the underlying GPU can queried via the `hipGetDeviceProperties` API and are stored in the `hipDeviceProp_t::gcnArchName` field.

This PR has 3 commits
1. Add the hooks to retrieve and query the `hipDeviceProp_t::gcnArchName` field
2. Update the `GpuVersion` datatype for AMDGPU in XLA code to include the `gcnArchName` string
3. parse the `gcnArchName` string and map the tokens within it, to appropriate "target-feature" string tokens.

The mapping done in the 3rd commit will be straight-forward once support for this feature gets upstreamed to the public LLVM repo. Until that happens and the TF LLVM pointer is moved beyond the upstream commit, we will need to special case the mapping. 

------------------------------------------------------------------------------------------

/cc @cheshire @chsigg @nvining-work

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc [hunk 1]
```diff
@@ -786,29 +786,82 @@ Status AMDGPUTargetModuleLinker(llvm::Module* module, GpuVersion gpu_version,
...
+  // For ROCM versions 3.9 and 3.10, hardcode it to empty string
+  feature_str = "";
+#else
+  // For ROCm versions 4.0 and greater, we need to specify the correct
+  // feature str, based on the underlying GPU WH to get max performance.
   feature_str = "";
+  int token_num = 0;
+  size_t start = 0, pos = 0;
+  do {
+    pos = gcn_arch_name.find(":", start);
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc:840]: Can we express this loop using more high-level APIs? (e.g. absl::StrSplit / absl::StrJoin?)

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc [hunk 2]
```diff
@@ -786,29 +786,82 @@ Status AMDGPUTargetModuleLinker(llvm::Module* module, GpuVersion gpu_version,
...
+// to be used for creating the AMDGPUTarget.
+// This mapping is currently in a state of flux because TF XLA uses its
+// own copy of LLVM, which is different from the LLVM version used by
+// hipcc/runtime in the ROCm install. Ordinarily this is not a problem,
+// but right now, the LLVM version used by hipcc/runtime has "targetID"
+// related changes which have not yet been upstreamed (to the LLVM repo)
+// When that upstreaming happens (and TF LLVM pointer moves past the
+// upstream commit), the following mapping will need to change
+static std::string MapGCNArchNameTokenToFeatureStr(const std::string token) {
+  if (token == "sramecc+")
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc:811]: Explicit braces are preferred even for one-liners

**Review by cheshire — APPROVED** [review]

**Comment by cheshire** [comment]
Seems to break some tests?

```
FAIL: //tensorflow/python/eager:function_test_gpu (shard 7 of 15) (see /tmpfs/bazel_output/_bazel_kbuilder/9a46bec5c8e82ceb5b4b229201d44d70/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/eager/function_test_gpu/shard_7_of_15/test_attempts/attempt_1.log)
```

Does this work on your end?

**Comment by cheshire** [comment]
```
terminate called without an active exception
Fatal Python error: Aborted

Thread 0x00007ff50d7b2700 (most recent call first):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/eager/function_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/c_api_util.py", line 58 in __del__

Thread 0x00007ff50ffb7700 (most recent call first):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/eager/function_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/c_api_util.py", line 58 in __del__

Thread 0x00007ff510fb9700 (most recent call first):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/eager/function_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/c_api_util.py", line 58 in __del__

Thread 0x00007ff511fbb700 (most recent call first):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/eager/function_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/c_api_util.py", line 58 in __del__

Current thread 0x00007ff5287e8700 (most recent call first):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/eager/function_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/c_api_util.py", line 58 in __del__

Thread 0x00007ff528fe9700 (most recent call first):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/eager/function_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/c_api_util.py", line 58 in __del__

Thread 0x00007ff5327fc700 (most recent call first):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/eager/function_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/c_api_util.py", line 58 in __del__

Thread 0x00007ff5397fa700 (most recent call first):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/eager/function_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/c_api_util.py", line 58 in __del__

Thread 0x00007ff53b7fe700 (most recent call first):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/eager/function_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/c_api_util.py", line 58 in __del__

Thread 0x00007ff5496a2700 (most recent call first):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/eager/function_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/c_api_util.py", line 58 in __del__

Thread 0x00007ff9357fa700 (most recent call first):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/eager/function_test_gpu.runfiles/org_tensorflow/tensorflow/python/framework/c_api_util.py", line 58 in __del__

```

**Comment by deven-amd** [comment]
> Seems to break some tests?
> 
> ```
> FAIL: //tensorflow/python/eager:function_test_gpu (shard 7 of 15) (see /tmpfs/bazel_output/_bazel_kbuilder/9a46bec5c8e82ceb5b4b229201d44d70/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/eager/function_test_gpu/shard_7_of_15/test_attempts/attempt_1.log)
> ```
> 
> Does this work on your end?

The changes in this PR do not (or atleast should not) have any impact on functionality...so doubtful that they are the cause of the regression.

Looking at the log (and assuming you are running internally on ROCm) , it seems that you are running with sharding enabled. Enabling sharding is not something we have had much success with internally...our CI script runs with sharding explicitly disabled!

Let me run that test locally and confirm it is working

**Comment by akuegel** [comment]
> > Seems to break some tests?
> > ```
> > FAIL: //tensorflow/python/eager:function_test_gpu (shard 7 of 15) (see /tmpfs/bazel_output/_bazel_kbuilder/9a46bec5c8e82ceb5b4b229201d44d70/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/eager/function_test_gpu/shard_7_of_15/test_attempts/attempt_1.log)
> > ```
> > 
> > 
> > Does this work on your end?
> 
> The changes in this PR do not (or atleast should not) have any impact on functionality...so doubtful that they are the cause of the regression.
> 
> Looking at the log (and assuming you are running internally on ROCm) , it seems that you are running with sharding enabled. Enabling sharding is not something we have had much success with internally...our CI script runs with sharding explicitly disabled!
> 
> Let me run that test locally and confirm it is working

The actual problem is that you forgot to update the nvptx_compiler.cc overload for GetGpuVersion. Also, some MLIR code needs to be adapted as well. I will fix it and try to get your change landed.

**Comment by deven-amd** [comment]
> Ok, I am trying to fix this now. There is some adjustment needed in gpu_kernel_to_blob_pass.cc (for the AMD specific code in kernel generator).

@akuegel , when I tried to do `bazel test //tensorflow/core/kernels/mlir_generated/...` locally, all 165 tests contained within it failed. 

The cause turned out be that `--arch=sm_70,compute_75` is hardcoded here (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/mlir_generated/build_test.sh#L27)

If I change the argument `--arch=gfx900` locally, all tests start passing.

Not sure whether the error you running into on your end is the same, or something different.

The above experiment was done on the `master` branch (i.e. not on the branch containing my changes). , going to try that next to see if there are any further failures caused by it.

Also please let me know if the error(s) you are seeing on your end are different, and if so how to go about reproducing them on my end

**Comment by deven-amd** [comment]
with the branch for this PR, I am getting an additional error
```
Execution platform: @local_execution_config_platform//:platform
tensorflow/compiler/mlir/tools/kernel_gen/transforms/gpu_kernel_to_blob_pass.cc: In member function 'xla::StatusOr<std::vector<unsigned char, std::allocator<unsigned char> > > mlir::kernel_gen::transforms::{anonymous}::GpuKernelToBlobPass::GetGpuBinaryBlob(mlir::gpu::GPUModuleOp)':
tensorflow/compiler/mlir/tools/kernel_gen/transforms/gpu_kernel_to_blob_pass.cc:124:62: error: could not convert 'arch' from 'uint32_t {aka unsigned int}' to 'xla::gpu::GpuVersion {aka absl::lts_2020_02_25::variant<std::pair<int, int>, std::pair<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >}'
           llvm_module_copy.get(), arch, config, libdevice_dir);
                                                              ^
tensorflow/compiler/mlir/tools/kernel_gen/transforms/gpu_kernel_to_blob_pass.cc:137:52: error: no matching function for call to 'std::vector<stream_executor::HsacoImage>::push_back(<brace-enclosed initializer list>)'
       images.push_back({arch_str, std::move(hsaco)});
                                                    ^
I
```

working on fixing it now

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc (hunk 1 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc:840]: done

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc (hunk 2 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc:811]: done

**Comment by deven-amd** [comment]
@cheshire @akuegel 

I have rebased the PR ( to the tip) and pushed out two more commits
1. to address the review comments from George
2. to address the build error in `gpu_kernel_to_blob_pass.cc` (potentially the same one that Adrian ran into) 

Please re-review and re-approve.

@akuegel,  I am assuming you will handle the issue of `--arch=sm_70,compute_75` being hard-coded in `.../build_test.sh` on your end.

**Comment by akuegel** [comment]
> @cheshire @akuegel
> 
> I have rebased the PR ( to the tip) and pushed out two more commits
> 
> 1. to address the review comments from George
> 2. to address the build error in `gpu_kernel_to_blob_pass.cc` (potentially the same one that Adrian ran into)
> 
> Please re-review and re-approve.
> 
> @akuegel, I am assuming you will handle the issue of `--arch=sm_70,compute_75` being hard-coded in `.../build_test.sh` on your end.

I was talking about the gpu_kernel_to_blob_pass.cc. The other thing is not an issue on our side, because we are running the tests only with Nvidia GPUs. Apart from gpu_kernel_to_blob_pass, I had fixed quite a few more things in your PR which ClangTidy pointed out, unfortunately the reviewer I asked for approval didn't react. I added another one now, hopefully I will get approval today.

**Review by akuegel — APPROVED** [review]
I think it might speed things up if I try to get your changes landed first, and then apply the additional ClangTidy warning fixes that I have prepared afterwards.

**Pull request merged**